src/theory/arith/arith_msum.cpp

```diff
@@ -0,0 +1,371 @@
+                  it->second.isNull()
+                      ? NodeManager::currentNM()->mkConst(Rational(-1))
+                      : negate(it->second);
+            }
+          }
+          return true;
+        }
+      }
+    }
+  }
+  return false;
+}
+
+Node ArithMSum::mkNode(const std::map<Node, Node>& msum)
+{
+  std::vector<Node> children;
+  for (std::map<Node, Node>::const_iterator it = msum.begin(); it != msum.end();
+       ++it)
+  {
+    Node m;
+    if (!it->first.isNull())
+    {
+      if (!it->second.isNull())
+      {
+        m = NodeManager::currentNM()->mkNode(MULT, it->second, it->first);
+      }
+      else
+      {
+        m = it->first;
+      }
+    }
+    else
+    {
+      Assert(!it->second.isNull());
+      m = it->second;
+    }
+    children.push_back(m);
+  }
+  return children.size() > 1
+             ? NodeManager::currentNM()->mkNode(PLUS, children)
+             : (children.size() == 1
+                    ? children[0]
+                    : NodeManager::currentNM()->mkConst(Rational(0)));
+}
+
+Node ArithMSum::mkCoeffTerm(Node coeff, Node t)
+{
+  if (coeff.isNull())
+  {
+    return t;
+  }
+  else
+  {
+    return NodeManager::currentNM()->mkNode(kind::MULT, coeff, t);
+  }
+}
+
+int ArithMSum::isolate(
+    Node v, const std::map<Node, Node>& msum, Node& veq_c, Node& val, Kind k)
+{
+  std::map<Node, Node>::const_iterator itv = msum.find(v);
+  if (itv != msum.end())
+  {
+    std::vector<Node> children;
+    Rational r =
+        itv->second.isNull() ? Rational(1) : itv->second.getConst<Rational>();
+    if (r.sgn() != 0)
+    {
+      for (std::map<Node, Node>::const_iterator it = msum.begin();
+           it != msum.end();
+           ++it)
+      {
+        if (it->first != v)
+        {
+          Node m;
+          if (!it->first.isNull())
+          {
+            if (!it->second.isNull())
+            {
+              m = NodeManager::currentNM()->mkNode(MULT, it->second, it->first);
+            }
+            else
+            {
+              m = it->first;
+            }
+          }
+          else
+          {
+            m = it->second;
+          }
+          children.push_back(m);
+        }
+      }
+      val = children.size() > 1
+                ? NodeManager::currentNM()->mkNode(PLUS, children)
+                : (children.size() == 1
+                       ? children[0]
+                       : NodeManager::currentNM()->mkConst(Rational(0)));
+      if (!r.isOne() && !r.isNegativeOne())
+      {
+        if (v.getType().isInteger())
+        {
+          veq_c = NodeManager::currentNM()->mkConst(r.abs());
+        }
+        else
+        {
+          val = NodeManager::currentNM()->mkNode(
+              MULT,
+              val,
+              NodeManager::currentNM()->mkConst(Rational(1) / r.abs()));
+        }
+      }
+      if (r.sgn() == 1)
+      {
+        val = negate(val);
+      }
+      else
+      {
+        val = Rewriter::rewrite(val);
+      }
```

Comment:
Again, imo easier to read if you use `val = ... ? ... : ... ;`

Reply:
I've made mkCoeffTerm inline and made these portions call that function.